src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs

```diff
@@ -76,27 +77,57 @@ public class HubEndPoint<THub> : IInvocationBinder where THub : Hub
             }
 
             connectionContext.UserIdentifier = _userIdProvider.GetUserId(connectionContext);
+            var protocolReaderWriter = connectionContext.ProtocolReaderWriter;
+
+            var needKeepAlive = connection.Features.Get<IConnectionInherentKeepAliveFeature>() == null;
 
             // Hubs support multiple producers so we set up this loop to copy
             // data written to the HubConnectionContext's channel to the transport channel
-            var protocolReaderWriter = connectionContext.ProtocolReaderWriter;
             async Task WriteToTransport()
             {
+                var serializedKeepAlive = protocolReaderWriter.WriteMessage(PingMessage.Instance);
+
                 try
                 {
-                    while (await output.Reader.WaitToReadAsync())
+                    while (true)
                     {
-                        while (output.Reader.TryRead(out var hubMessage))
+                        var keepAliveTask = needKeepAlive ?
+                            Task.Delay(_hubOptions.KeepAliveInterval) :
```

Comment:
Should I be concerned that we new up a Task.Delay every send?

Reply:
Not every send. Every time we need to go async to wait for the user to send. And you have to allocate a task for this no matter how you slice it.